### PR TITLE
Explicitly define metric reader force flush operation

### PR DESCRIPTION
**Pull request by legendecas**
Related: https://github.com/open-telemetry/opentelemetry-specification/issues/2983
Related: https://github.com/open-telemetry/opentelemetry-js/issues/3278

## Changes

> ForceFlush MUST invoke ForceFlush on all registered [MetricReader](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#metricreader) and [Push Metric Exporter](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#push-metric-exporter) instances.

The method `MetricReader.ForceFlush` is already mentioned in https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#forceflush. However, the specification doesn't explicitly define the method `MetricReader.ForceFlush`.

This also suggests that `ForceFlush` of Periodic exporting MetricReader should call `Collect` to consume the metrics.

### CODE REVIEW

**Review by MrAlias — APPROVED** [review]

File: specification/metrics/sdk.md [hunk 1]
```diff
@@ -882,6 +879,23 @@ Note: it is expected that the `MetricReader.Collect` implementations will be
...
+`ForceFlush` SHOULD provide a way to let the caller know whether it succeeded,
+failed or timed out. Note: unlike [Push Metric Exporter](#push-metric-exporter)
+which can send data on its own schedule,
+[Pull Metric Exporter](#pull-metric-exporter) can only send the data when it is
+being asked by the scraper, so `ForceFlush` would not make much sense.
+
+`ForceFlush` SHOULD complete or abort within some timeout. `ForceFlush` MAY be
+implemented as a blocking API or an asynchronous API which notifies the caller
+via a callback or an event. [OpenTelemetry SDK](../overview.md#sdk) authors MAY
+decide if they want to make the shutdown timeout configurable.
```
Comment (reyang) [inline specification/metrics/sdk.md:897]:
```suggestion
decide if they want to make the timeout configurable.
```

**Comment by reyang** [comment]
> The method `MetricReader.ForceFlush` is already mentioned in https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#forceflush. However, the specification doesn't explicitly define the method `MetricReader.ForceFlush`.
> 
> This also suggests that `ForceFlush` of Periodic exporting MetricReader should call `Collect` to consume the metrics.

Could this be fixed by changing:

ForceFlush MUST invoke ForceFlush on all registered [MetricReader](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#metricreader) and [Push Metric Exporter](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#push-metric-exporter) instances.

to:

ForceFlush MUST invoke `Collect` on all registered [MetricReader](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#metricreader) and `ForceFlush` on all [Push Metric Exporter](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#push-metric-exporter) instances.

**Comment by legendecas** [comment]
@reyang for pull metric reader / metric exporter, I don't think it is meaningful to invoke `Collect` on force flush, like mentioned at https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#pull-metric-exporter.

**Comment by legendecas** [comment]
Not stale, friendly ping @reyang

**Comment by reyang** [comment]
> Not stale, friendly ping @reyang

Hi @legendecas, what's expected from me?

**Comment by legendecas** [comment]
@reyang has your question https://github.com/open-telemetry/opentelemetry-specification/pull/3084#issuecomment-1380784228 been addressed with https://github.com/open-telemetry/opentelemetry-specification/pull/3084#issuecomment-1381354877? If so, is this PR still looking good to you?

**Comment by reyang** [comment]
> @reyang has your question [#3084 (comment)](https://github.com/open-telemetry/opentelemetry-specification/pull/3084#issuecomment-1380784228) been addressed with [#3084 (comment)](https://github.com/open-telemetry/opentelemetry-specification/pull/3084#issuecomment-1381354877)? If so, is this PR still looking good to you?

I didn't ask a question, in the comment I had a suggestion. With the change proposed by the current PR, I find it hard to understand what's the difference between MetricReader.Collect and MetricReader.ForceFlush.

**Comment by legendecas** [comment]
@reyang `MetricReader.Collect` can be initiated by external collectors like prometheus scraping. However, `MetricReader.ForceFlush` is intended for sdk to notify MetricReaders that they should consume and send as much metric as they can.

> `ForceFlush` MUST invoke `ForceFlush` on all registered
[MetricReader](#metricreader) and [Push Metric Exporter](#push-metric-exporter)
instances.

The statement in the spec has already mentioned that SDK's `ForceFlush` should invoke `MetricReader`'s `ForceFlush`. This change doesn't change this behavior.

**Pull request closed without merging**